File: src/precalculus_unit_circle_easy.py
"""
precalculus - unit_circle (easy)
"""

from problem_utils import *

# All 16 standard angles (every 22.5° is non-exact, so stick to 30-family + 45-family)
ANGLES_DEG = [0, 30, 45, 60, 90, 120, 135, 150, 180,
              210, 225, 240, 270, 300, 315, 330, 360]
SAFE_TAN   = [d for d in ANGLES_DEG if d not in (90, 270, 360)]

# Extended Pythagorean triples for "find the other trig value" problems
TRIPLES = [(3,4,5), (5,12,13), (8,15,17), (7,24,25), (20,21,29), (9,40,41)]


def generate():
    problem_type = randint(1, 5)

    if problem_type == 1:
        # Find exact value of sin/cos/tan
        func_choice = choice(['sin', 'cos', 'tan'])
        if func_choice == 'tan':
            angle_deg = choice(SAFE_TAN)
        else:
            angle_deg = choice(ANGLES_DEG)

        angle_rad = angle_deg * pi / 180
        if func_choice == 'sin':
            ans = simplify(sin(angle_rad))
            func_name = '\\sin'
        elif func_choice == 'cos':
            ans = simplify(cos(angle_rad))
            func_name = '\\cos'
        else:
            ans = simplify(tan(angle_rad))
            func_name = '\\tan'

        return problem(
            question=f"Find the exact value: ${func_name}\\!\\left({latex(angle_rad)}\\right)$",
            answer=ans,
            difficulty=(1000, 1100),
            topic="precalculus/unit_circle",
            solution=steps(
                f"The angle ${latex(angle_rad)}$ rad $= {angle_deg}^\\circ$",
                f"Using the unit circle: ${func_name}({angle_deg}^\\circ) = {latex(ans)}$"
            )
        )

    elif problem_type == 2:
        # Convert between degrees and radians
        convert_to = choice(['radians', 'degrees'])
        if convert_to == 'radians':
            angle_deg = choice([d for d in ANGLES_DEG if d != 0])
            ans = simplify(angle_deg * pi / 180)
            return problem(
                question=f"Convert ${angle_deg}^\\circ$ to radians.",
                answer=ans,
                difficulty=(1000, 1150),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"Multiply by $\\dfrac{{\\pi}}{{180}}$: ${angle_deg} \\times \\dfrac{{\\pi}}{{180}} = {latex(ans)}$"
                )
            )
        else:
            common_rads = [pi/6, pi/4, pi/3, pi/2, 2*pi/3, 3*pi/4, 5*pi/6, pi,
                           7*pi/6, 5*pi/4, 4*pi/3, 3*pi/2, 5*pi/3, 7*pi/4, 11*pi/6, 2*pi]
            angle_rad = choice(common_rads)
            ans = simplify(angle_rad * 180 / pi)
            return problem(
                question=f"Convert ${latex(angle_rad)}$ radians to degrees.",
                answer=ans,
                difficulty=(1000, 1150),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"Multiply by $\\dfrac{{180}}{{\\pi}}$: ${latex(angle_rad)} \\times \\dfrac{{180}}{{\\pi}} = {latex(ans)}^\\circ$"
                )
            )

    elif problem_type == 3:
        # Find the reference angle — varied quadrants
        angle_deg = choice([d for d in ANGLES_DEG if d not in (0, 90, 180, 270, 360)])

        if 90 < angle_deg <= 180:
            ref = 180 - angle_deg
            quad = "II"
        elif 180 < angle_deg <= 270:
            ref = angle_deg - 180
            quad = "III"
        else:
            ref = 360 - angle_deg
            quad = "IV"

        return problem(
            question=f"Find the reference angle for ${angle_deg}^\\circ$.",
            answer=ref,
            difficulty=(1100, 1200),
            topic="precalculus/unit_circle",
            solution=steps(
                f"${angle_deg}^\\circ$ is in quadrant {quad}",
                f"Reference angle $= {ref}^\\circ$"
            )
        )

    elif problem_type == 4:
        # Find (x, y) coordinates on the unit circle
        angle_deg = choice(ANGLES_DEG)
        angle_rad = angle_deg * pi / 180
        x_coord = simplify(cos(angle_rad))
        y_coord = simplify(sin(angle_rad))

        # Randomly ask for x, y, or both
        ask = choice(['x', 'y', 'both'])
        if ask == 'x':
            return problem(
                question=f"On the unit circle at angle ${latex(angle_rad)}$ rad, what is the $x$-coordinate?",
                answer=x_coord,
                difficulty=(1100, 1250),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"$x = \\cos\\theta = \\cos({angle_deg}^\\circ) = {latex(x_coord)}$"
                )
            )
        elif ask == 'y':
            return problem(
                question=f"On the unit circle at angle ${latex(angle_rad)}$ rad, what is the $y$-coordinate?",
                answer=y_coord,
                difficulty=(1100, 1250),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"$y = \\sin\\theta = \\sin({angle_deg}^\\circ) = {latex(y_coord)}$"
                )
            )
        else:
            return problem(
                question=f"Find the coordinates $(x, y)$ on the unit circle at ${latex(angle_rad)}$ radians.",
                answer=(x_coord, y_coord),
                difficulty=(1100, 1250),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"$x = \\cos({angle_deg}^\\circ) = {latex(x_coord)}$",
                    f"$y = \\sin({angle_deg}^\\circ) = {latex(y_coord)}$"
                )
            )

    else:
        # Given one trig value and quadrant, find another using Pythagorean identity
        triple = choice(TRIPLES)
        a_val, b_val, c_val = triple
        given_func = choice(['sin', 'cos'])
        quadrant = choice([1, 2, 3, 4])

        if given_func == 'sin':
            given_val = Rational(a_val, c_val)
            # cos^2 = 1 - sin^2 = b^2/c^2
            cos_mag = Rational(b_val, c_val)
            if quadrant in (1, 4):
                ans = cos_mag
                sign_str = "positive"
            else:
                ans = -cos_mag
                sign_str = "negative"
            return problem(
                question=f"If $\\sin\\theta = {latex(given_val)}$ and $\\theta$ is in quadrant {quadrant}, find $\\cos\\theta$.",
                answer=ans,
                difficulty=(1200, 1300),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"$\\cos^2\\theta = 1 - \\sin^2\\theta = 1 - {latex(given_val**2)} = {latex(cos_mag**2)}$",
                    f"$\\cos\\theta = \\pm{latex(cos_mag)}$",
                    f"Quadrant {quadrant}: cosine is {sign_str}",
                    f"$\\cos\\theta = {latex(ans)}$"
                )
            )
        else:
            given_val = Rational(b_val, c_val)
            sin_mag = Rational(a_val, c_val)
            if quadrant in (1, 2):
                ans = sin_mag
                sign_str = "positive"
            else:
                ans = -sin_mag
                sign_str = "negative"
            return problem(
                question=f"If $\\cos\\theta = {latex(given_val)}$ and $\\theta$ is in quadrant {quadrant}, find $\\sin\\theta$.",
                answer=ans,
                difficulty=(1200, 1300),
                topic="precalculus/unit_circle",
                solution=steps(
                    f"$\\sin^2\\theta = 1 - \\cos^2\\theta = 1 - {latex(given_val**2)} = {latex(sin_mag**2)}$",
                    f"$\\sin\\theta = \\pm{latex(sin_mag)}$",
                    f"Quadrant {quadrant}: sine is {sign_str}",
                    f"$\\sin\\theta = {latex(ans)}$"
                )
            )

emit(generate())

File: src/precalculus_unit_circle_easy.jl
# precalculus - unit_circle (easy)
# Generated: 2026-03-08T20:27:22.776203

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script θ begin
    set_topic!("precalculus/unit_circle")
    
    # Choose problem type randomly
    problem_type = choice(1:5)
    
    if problem_type == 1
        # Find exact trig value at standard angle (700-900 ELO)
        angles_deg = [0, 30, 45, 60, 90, 120, 135, 150, 180, 210, 225, 240, 270, 300, 315, 330]
        angle = choice(angles_deg)
        func = choice(["sin", "cos", "tan"])
        
        # Calculate exact answer
        angle_rad = angle * π / 180
        if func == "sin"
            if angle == 0 || angle == 180
                ans = 0
            elseif angle == 30 || angle == 150
                ans = 1//2
            elseif angle == 45 || angle == 135
                ans = "\\frac{\\sqrt{2}}{2}"
            elseif angle == 60 || angle == 120
                ans = "\\frac{\\sqrt{3}}{2}"
            elseif angle == 90
                ans = 1
            elseif angle == 210 || angle == 330
                ans = -1//2
            elseif angle == 225 || angle == 315
                ans = "-\\frac{\\sqrt{2}}{2}"
            elseif angle == 240 || angle == 300
                ans = "-\\frac{\\sqrt{3}}{2}"
            elseif angle == 270
                ans = -1
            end
        elseif func == "cos"
            if angle == 0
                ans = 1
            elseif angle == 30 || angle == 330
                ans = "\\frac{\\sqrt{3}}{2}"
            elseif angle == 45 || angle == 315
                ans = "\\frac{\\sqrt{2}}{2}"
            elseif angle == 60 || angle == 300
                ans = 1//2
            elseif angle == 90 || angle == 270
                ans = 0
            elseif angle == 120 || angle == 240
                ans = -1//2
            elseif angle == 135 || angle == 225
                ans = "-\\frac{\\sqrt{2}}{2}"
            elseif angle == 150 || angle == 210
                ans = "-\\frac{\\sqrt{3}}{2}"
            elseif angle == 180
                ans = -1
            end
        else  # tan
            if angle == 0 || angle == 180
                ans = 0
            elseif angle == 30 || angle == 210
                ans = "\\frac{\\sqrt{3}}{3}"
            elseif angle == 45 || angle == 225
                ans = 1
            elseif angle == 60 || angle == 240
                ans = "\\sqrt{3}"
            elseif angle == 120 || angle == 300
                ans = "-\\sqrt{3}"
            elseif angle == 135 || angle == 315
                ans = -1
            elseif angle == 150 || angle == 330
                ans = "-\\frac{\\sqrt{3}}{3}"
            else
                ans = "\\text{undefined}"
            end
        end
        
        problem(
            question="Find the exact value: \\\$\\\$(func)($(angle)^\\circ)\\\$",
            answer=ans,
            difficulty=(700, 900),
            solution=steps(
                "Locate \\\$$(angle)^\\circ\\\$ on the unit circle",
                "Identify the $(func == "sin" ? "y" : func == "cos" ? "x" : "y/x")-coordinate$(func == "tan" ? " ratio" : "")",
                sol("Answer", ans)
            ),
            time=45
        )
        
    elseif problem_type == 2
        # Convert between radians and degrees (700-900 ELO)
        direction = choice(["deg_to_rad", "rad_to_deg"])
        
        if direction == "deg_to_rad"
            deg_values = [30, 45, 60, 90, 120, 135, 150, 180, 210, 225, 240, 270, 300, 315, 330, 360]
            deg = choice(deg_values)
            
            # Simplify to nice fraction
            if deg == 30
                ans = "\\frac{\\pi}{6}"
            elseif deg == 45
                ans = "\\frac{\\pi}{4}"
            elseif deg == 60
                ans = "\\frac{\\pi}{3}"
            elseif deg == 90
                ans = "\\frac{\\pi}{2}"
            elseif deg == 120
                ans = "\\frac{2\\pi}{3}"
            elseif deg == 135
                ans = "\\frac{3\\pi}{4}"
            elseif deg == 150
                ans = "\\frac{5\\pi}{6}"
            elseif deg == 180
                ans = "\\pi"
            elseif deg == 210
                ans = "\\frac{7\\pi}{6}"
            elseif deg == 225
                ans = "\\frac{5\\pi}{4}"
            elseif deg == 240
                ans = "\\frac{4\\pi}{3}"
            elseif deg == 270
                ans = "\\frac{3\\pi}{2}"
            elseif deg == 300
                ans = "\\frac{5\\pi}{3}"
            elseif deg == 315
                ans = "\\frac{7\\pi}{4}"
            elseif deg == 330
                ans = "\\frac{11\\pi}{6}"
            elseif deg == 360
                ans = "2\\pi"
            end
            
            problem(
                question="Convert \\\$$(deg)^\\circ\\\$ to radians.",
                answer=ans,
                difficulty=(700, 850),
                solution=steps(
                    "Use the conversion: \\\$\\text{radians} = \\text{degrees} \\times \\frac{\\pi}{180}\\\$",
                    "Calculate: \\\$$(deg) \\times \\frac{\\pi}{180}\\\$",
                    sol("Answer", ans)
                ),
                time=60
            )
        else
            rad_values = [(1,6), (1,4), (1,3), (1,2), (2,3), (3,4), (5,6), (1,1), (7,6), (5,4), (4,3), (3,2), (5,3), (7,4), (11,6), (2,1)]
            num, den = choice(rad_values)
            
            deg = 180 * num // den
            
            problem(
                question="Convert \\\$\\frac{$(num)\\pi}{$(den)}\\\$ radians to degrees.",
                answer=deg,
                difficulty=(700, 850),
                solution=steps(
                    "Use the conversion: \\\$\\text{degrees} = \\text{radians} \\times \\frac{180}{\\pi}\\\$",
                    "Calculate: \\\$\\frac{$(num)\\pi}{$(den)} \\times \\frac{180}{\\pi}\\\$",
                    sol("Answer", "$(deg)^\\circ")
                ),
                time=60
            )
        end
        
    elseif problem_type == 3
        # Find angle given trig value (800-1000 ELO)
        func = choice(["sin", "cos"])
        quadrant = choice([1, 2, 3, 4])
        
        ref_angles = [30, 45, 60]
        ref = choice(ref_angles)
        
        if func == "sin"
            if ref == 30
                val = "\\frac{1}{2}"
            elseif ref == 45
                val = "\\frac{\\sqrt{2}}{2}"
            else
                val = "\\frac{\\sqrt{3}}{2}"
            end
            
            if quadrant == 1
                angle = ref
            elseif quadrant == 2
                angle = 180 - ref
            elseif quadrant == 3
                angle = 180 + ref
                val = "-" * val
            else
                angle = 360 - ref
                val = "-" * val
            end
        else  # cos
            if ref == 30
                val = "\\frac{\\sqrt{3}}{2}"
            elseif ref == 45
                val = "\\frac{\\sqrt{2}}{2}"
            else
                val = "\\frac{1}{2}"
            end
            
            if quadrant == 1
                angle = ref
            elseif quadrant == 2
                angle = 180 - ref
                val = "-" * val
            elseif quadrant == 3
                angle = 180 + ref
                val = "-" * val
            else
                angle = 360 - ref
            end
        end
        
        problem(
            question="Find the angle \\\$\\theta\\\$ in \\\$[0^\\circ, 360^\\circ)\\\$ where \\\$\\\$(func)(\\theta) = $(val)\\\$ and \\\$\\theta\\\$ is in quadrant $(quadrant).",
            answer=angle,
            difficulty=(900, 1100),
            solution=steps(
                "Reference angle: \\\$$(ref)^\\circ\\\$ (since \\\$|$(func)(\\theta)| = $(replace(val, "-" => ""))\\\$)",
                "In quadrant $(quadrant), adjust the reference angle",
                sol("Answer", "$(angle)^\\circ")
            ),
            time=90
        )
        
    elseif problem_type == 4
        # Find coordinate on unit circle (800-1000 ELO)
        angles_deg = [30, 45, 60, 120, 135, 150, 210, 225, 240, 300, 315, 330]
        angle = choice(angles_deg)
        
        # Calculate coordinates
        if angle == 30
            coords = "\\left(\\frac{\\sqrt{3}}{2}, \\frac{1}{2}\\right)"
        elseif angle == 45
            coords = "\\left(\\frac{\\sqrt{2}}{2}, \\frac{\\sqrt{2}}{2}\\right)"
        elseif angle == 60
            coords = "\\left(\\frac{1}{2}, \\frac{\\sqrt{3}}{2}\\right)"
        elseif angle == 120
            coords = "\\left(-\\frac{1}{2}, \\frac{\\sqrt{3}}{2}\\right)"
        elseif angle == 135
            coords = "\\left(-\\frac{\\sqrt{2}}{2}, \\frac{\\sqrt{2}}{2}\\right)"
        elseif angle == 150
            coords = "\\left(-\\frac{\\sqrt{3}}{2}, \\frac{1}{2}\\right)"
        elseif angle == 210
            coords = "\\left(-\\frac{\\sqrt{3}}{2}, -\\frac{1}{2}\\right)"
        elseif angle == 225
            coords = "\\left(-\\frac{\\sqrt{2}}{2}, -\\frac{\\sqrt{2}}{2}\\right)"
        elseif angle == 240
            coords = "\\left(-\\frac{1}{2}, -\\frac{\\sqrt{3}}{2}\\right)"
        elseif angle == 300
            coords = "\\left(\\frac{1}{2}, -\\frac{\\sqrt{3}}{2}\\right)"
        elseif angle == 315
            coords = "\\left(\\frac{\\sqrt{2}}{2}, -\\frac{\\sqrt{2}}{2}\\right)"
        elseif angle == 330
            coords = "\\left(\\frac{\\sqrt{3}}{2}, -\\frac{1}{2}\\right)"
        end
        
        problem(
            question="Find the coordinates \$(x, y)\\\$ of the point on the unit circle at angle \\\$$(angle)^\\circ\\\$.",
            answer=coords,
            difficulty=(850, 1050),
            solution=steps(
                "On the unit circle, \\\$x = \\cos(\\theta)\\\$ and \\\$y = \\sin(\\theta)\\\$",
                "Calculate \\\$\\cos($(angle)^\\circ)\\\$ and \\\$\\sin($(angle)^\\circ)\\\$",
                sol("Answer", coords)
            ),
            time=75
        )
        
    else  # problem_type == 5
        # Solve simple trig equation (1000-1200 ELO)
        func = choice(["sin", "cos"])
        
        ref_angles = [30, 45, 60]
        ref = choice(ref_angles)
        
        if func == "sin"
            if ref == 30
                val = 1//2
                val_str = "\\frac{1}{2}"
            elseif ref == 45
                val = "sqrt2/2"
                val_str = "\\frac{\\sqrt{2}}{2}"
            else
                val = "sqrt3/2"
                val_str = "\\frac{\\sqrt{3}}{2}"
            end
            
            # sin is positive in Q1 and Q2
            angle1 = ref
            angle2 = 180 - ref
        else  # cos
            if ref == 30
                val = "sqrt3/2"
                val_str = "\\frac{\\sqrt{3}}{2}"
            elseif ref == 45
                val = "sqrt2/2"
                val_str = "\\frac{\\sqrt{2}}{2}"
            else
                val = 1//2
                val_str = "\\frac{1}{2}"
            end
            
            # cos is positive in Q1 and Q4
            angle1 = ref
            angle2 = 360 - ref
        end
        
        ans = fmt_set(Set([angle1, angle2]))
        
        problem(
            question="Solve for \\\$\\theta\\\$ in \\\$[0^\\circ, 360^\\circ)\\\$: \\\$\\\$(func)(\\theta) = $(val_str)\\\$",
            answer=ans,
            difficulty=(1000, 1200),
            answer_type="set",
            solution=steps(
                "Reference angle is \\\$$(ref)^\\circ\\\$",
                "$(func) is positive in $(func == "sin" ? "quadrants I and II" : "quadrants I and IV")",
                "Solutions: \\\$\\theta = $(angle1)^\\circ\\\$ and \\\$\\theta = $(angle2)^\\circ\\\$",
                sol("Answer", "\\{$(angle1)^\\circ, $(angle2)^\\circ\\}")
            ),
            time=120
        )
    end
end